### DISCUSSION TITLE
Unrealistic stranding

### DISCUSSION THREAD

**Discussion by brucelaughlin** [OpenDrift/opendrift#1459] [General]
I am seeding particles representing fish larvae near a coastline, and many of my "fish" appear to become "stranded" at the coast, where model velocities go to zero.  I have been advised that this is unrealistic behavior, and a colleague suggested that I try adding an artificial "kick" to such stranded particles.

Has this issue come up for other users?  I may need to write the code for these "kicks" myself, so I'm wondering if any work has already been done on this or if any suggestions can be offered.  My model does not include tides.

Thank you.

**Comment by knutfrode** [comment]
Normally ocean currents are extrapolated towards coast, so a bit strange that you experience this.
Did you make a custom module, or one of the existing?
Are you using the default GSHHS coastline, or do you set configuration to use custom provided landmask?
And do these elements get deactivated as "stranded", or are they still active but just not moving?

**Reply by brucelaughlin** [reply]
"ocean currents are extrapolated towards coast" - does this mean that the interpolation which yields a velocity at a particle's location should not be zero, even if the particle is closer to a land grid point than an ocean point?

I essentially copied "oceandrift.py", so my "update()" method is the same.  

As for the coastline, I recently discovered that I had been using the GSHHS coastline, as I noticed particles entering bays which aren't in the model. I then discovered the "self._set_config_default('general:use_auto_landmask', False)" option which seemed to fix this problem.  In the beginning of the log, I see "INFO    opendrift.models.basemodel:935: Using existing reader for land_binary_mask", and then I periodically see "WARNING opendrift.models.basemodel:2327: Missing variables: ['land_binary_mask']", and an accompanying increase in the number of deactivated particles.  Perhaps this means a particle has hit the edge of the grid and become stranded there?  Should I be confident that my own mask is now being used, and that there isn't a missing variable?

Finally, these particles do NOT get deactivated as "stranded" - they are active.  Also, I am not 100% sure they are truly "stranded"... they may simply not be moving very much.  I have tried to reproduce this behavior in small, plot-able runs, but it only seems to appear when I seed throughout my entire experiment domain and then look at settlement statistics for thousands/millions of particles.

I'll keep looking for bugs...

**Comment by knutfrode** [comment]
Since you have deactivated the GSHHG landmask, `land_binary_mask` must be provided by one of your readers instead.
If you make a plot, you should be able to see whether the deactivated particles due to missing landmask are on the border of your model grid, or if it is towards coast.
It is difficult to debug by looking at post processing statistics, it might be best if you could reproduce the problem with a minimalistic test run.

**Reply by brucelaughlin** [reply]
Thank you, I'll look for this.  I do recall that these particles are along the boundaries and not along the coast.

Also, let me refine the problem statement.  I will be working on a solution, but I'm curious if my concerns seem valid:

What's happening, I think, is that particles near the coast (ie closer to the coast than to an offshore u/v-grid point), are being given a near-zero normal velocity as a product of linear interpolation.  So, for a north-south coastline, particles that manage to get close to the coast get "stuck" close to the coast, as the u-velocity value is zero along the coast and thus a linear interpolation of u-velocities gives these particles a near-zero u-velocity.  They may have a nonzero v-velocity and thus might travel up/down the coastline, but they may never again gain a non-zero u-velocity and will thus never "escape" the nearshore zone.  And, if particles enter a "corner" in the coastline, they will be prescribed near-zero u AND v-velocity values, by the same reasoning.  So, I believe that particles can get "artificially stranded" in this way.  A former colleague said that this behavior is unphysical (at least for fish larvae), and that they implemented random "kicks" to deal with this in the tracking software they were using.

**Comment by knutfrode** [comment]
Are you using the coastline option "previous" ? If so, particles hitting coast are moved back to the last position in water, and held there until currents turn and move them offshore again.
It might sound like you are using a quite large computation time step so that particles moved far back from coastline?
Calculation timestep should normally not be larger than 1 hour - but output time step may still be larger. For some application you might need time step of 15 minutes or less.

Recently, this "previous" option was improved (by TheSylex) by moving particles closer to the actual ocean-land-boundary. 
Thus you might see an improvement if you update OpenDrift to the latest version.
